backend/services/grading_embedding_service.py

```
@@ -0,0 +1,388 @@
+            else:
+                current_chunk += '. ' + sentence if current_chunk else sentence
+        
+        if current_chunk.strip():
+            chunks.append(current_chunk.strip())
+        
+        return chunks
+    
+    async def _get_embedding_async(self, text: str, cache_key: str) -> List[float]:
+        """Get embedding for text with caching"""
+        # Create cache key from content hash
+        content_hash = hashlib.md5(text.encode()).hexdigest()
```

Ruff (0.14.1)

Probable use of insecure hash functions in `hashlib`: `md5`

(S324)